Use tween when flipping

wontheone1 opened the pull request:
Use tween when flipping

Commit:
Buggy! Use tween when flipping

When button is clicked multiple times too fast, checking x scale is 0 is not correct because x might be not 0
yet but it can be in the middle of becoming 0

wontheone1 committed:
Use Phaser constants instead of string literal for Linear

wontheone1 committed:
Fix bug! Toggle by checking flipped state, not by checking scale is 0

flyjwayur commented on src/cljs/hello_world/game.cljs:
```diff
@@ -200,6 +201,8 @@
     (if (not (empty? initial-sprites-state))
       (util/synchronize-puzzle-board initial-sprites-state)
       (do
+        (swap! util/game-state assoc :sprites-state {})     ; prevent :sprites-state is nil when creating
```
I also saw the 'nill' error in console. Why it can be 'null', eventhough the game-state is assigned {} in util.cljs?:)

wontheone1 replied:
@flyjwayur On game start, server can send sprites-state as `nil`. look
https://github.com/wontheone1/aikakone/blob/master/src/clj/user.clj#L9
https://github.com/wontheone1/aikakone/blob/master/src/clj/hello_world/core.clj#L23
https://github.com/wontheone1/aikakone/blob/master/src/clj/hello_world/core.clj#L58

wontheone1 committed:
Extract create-puzzle-piece-and-store function

wontheone1 committed:
Button flipping converges to the center of pieces

wontheone1 committed:
Re-center the puzzle board again

wontheone1 commented:
@flyjwayur I have an idea to solve the bug that sometimes the puzzle is unsolvable state. 
Currently, we store the state piece by piece and sometimes a few pieces are not in sync with the state of row/col. I'll change so that we store flipped states of row/col/diagonal, and also only exchange states of row/col/diagonal through web-socket. And synchronizing each puzzle pieces to new states will happen in synchronized fashion I think this could solve the bug once and for all.

wontheone1 committed:
Switch position of row, col. Always row then col to not confused

wontheone1 committed:
Define sprites-state in terms of button-flipped, not in terms of each piece is flipped

(tween not applied yet)
Prevent some pieces are not in sync with their row/col/diagonal states.

wontheone1 committed:
Use tween when syncing puzzle board

wontheone1 committed:
Remove unnecessary js/setTimeout

When state is defined by row/col/diagonal flipped states,
it seems that pieces are always synchronized correctly so there is no need to wait for pieces to sync

wontheone1 commented:
Everything looks good now! Tween only executes on active browser,
so to see if the other browsers are in sync, you have to select that browser
then tween should execute.

flyjwayur commented:
@wontheone1 Thanks a lot for the excellent idea and solution. One thing I concerned is that players might think they can flip the certain row/col/digonal pieces which are already solved by other player,  before they activate the browser. They need to click somewhere in their browser to sync, so it might not bring wrong guess to flip and takes more time.

flyjwayur reviewed (commented):
@wontheone1 Thanks a lot for the brilliant idea and solution. It will solve the unsolvable state issue very well. One thing I concerned is that players might think they can flip the certain row/col/diagonal pieces which were already solved by other player, before they activate the browser(cuz it didnt sync the puzzle from the last player's flip) They need to click somewhere in their browser to sync, so it might bring wrong guess to flip and be inconvenient. Or it could be just my concern? :)

wontheone1 committed:
Initialize sprites-state from server

flyjwayur commented:
Thanks a lot for the brilliant ideas! 👍  :)

wontheone1 commented:
@flyjwayur as we discussed I think clients will be in sync if the browser is active. Let's see after we deploy it to some server. If they don't sync, we can remove tween.
